pkg/util/mount/mount_linux.go

```diff
@@ -1155,3 +1155,11 @@ func doSafeOpen(pathname string, base string) (int, error) {
 
 	return finalFD, nil
 }
+
+func (mounter *Mounter) GetAbsoluteHostPath(pathname string) (string, error) {
+	return doGetAbsoluteHostPath(pathname)
+}
+
+func doGetAbsoluteHostPath(pathname string) (string, error) {
+	return filepath.EvalSymlinks(pathname)
```

Comment:
* host-path volume: `/tmp/foo`
* subpath: `/tmp/foo/bar`
* `/tmp/foo/bar` is symlink to `/tmp/foo/baz` (using absolute path), ie. safe to be used.

Now, if kubelet runs inside a container, `/rootfs/tmp/foo/bar` gets resolved into `/tmp/foo/baz` which is NOT safe to be used and points to a very unexpected location.

I need to think about it.

Reply:
Agree this approach isn't going to work at all.  Especially because the proc path cannot be resolved on the host

Reply:
@jsafrane @msau42 How about writing a dedicated function to evaluate symlinks in containerized kubelet? e.g. `func EvalSymlinksAtRoot(path string, root string) (string, error)`.

IMO, `/`  in host is mounted in containerized kubelet "/rootfs". We can think it as a filesystem at non-root ("/") path. Any reading on path under "/rootfs", we can simply prepend "/rootfs" on path. 

`filepath.EvalSymlinks` does not work, because it assumes filesystem root is "/" internally (it uses`os.Lstat`, `os.Readlink` functions, etc). We can write our own function to evaluate symbolic links at non-root filesystem.

I modifed `filepath.EvalSymlinks` implementation, and prepend "/rootfs" before passing path to os functions (`os.Lstat`, `os.Readlink`, etc). Here is demonstration: https://gist.github.com/cofyc/860b134c593c1d81e19ca34a5c26bb6f.

Reply:
I think it is simpler to call readlink on the host and then prepend "/rootfs" to it, rather than:

1. prepend "/rootfs" to path
2. resolve symlinks in container
3. figure out if "/rootfs" prefix still exists
4. prepend "/rootfs" if it got dropped

Reply:
Using `EvalSymlinksAtRoot` don't need to figure out if "/rootfs" prefix still exists. In this case, `filepath.Join("/rootfs", EvalSymlinksAtRoot("/tmp/foo/bar", "/rootfs"))` will do.

`EvalSymlinksAtRoot` evaluates symbolic links in specified filesystem under given root (fs mountpoint). `EvalSymlinks(path)` equals to `EvalSymlinksAtRoot(path, "/")`.

I have done [some tests in container](https://gist.github.com/cofyc/860b134c593c1d81e19ca34a5c26bb6f#file-evalsymlinks_test-go), it works very well. I prefer this way if there is no special cases that cannot be avoided. Calling `readlink` on the host is simpler, but it needs to call `nsenter`,  then `readlink` commands, which as I think is too expensive way to resolve a path.

Reply:
We have to call nsenter when doing the bind mount too. I'm not sure if this extra complexity will save us much overall.

Reply:
Bind mount requires `nsenter`, which we cannot avoid. But to simply get a resolved path containerized kubelet can access (e.g. call `os.Lstat` in `volume.IsSameGroup` or check path exists or not), if we can avoid invoking external commands, that will be better, IMO.